### PR TITLE
Espresso SDK

### PR DESCRIPTION
**Pull request by kb-kerem**
Open points:

- ~~Environment variable handling needs to be adapted to Android best practices~~ Handled it with buildConfig and localProperties
- ~~Builder classes where appropriate (VisualClient, VisualCheckOptions, BuildAttributes etc.)~~ Provided
- ~~Use default Android http client instead of Apache's~~ Switched to using apollo graphql client

### CODE REVIEW

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/exception/VisualApiException.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.saucelabs.visual.exception;
+
+public class VisualApiException extends RuntimeException {
```
Comment (konraddysput) [inline visual-espresso/visual/src/main/java/com/saucelabs/visual/exception/VisualApiException.java:3]: Runtime Exception is a special exception - are you sure you want to extend the Runtime exception?  Throwing such an exception should kill the test suite. I think we want the test to fail instead?
```suggestion
public class VisualApiException extends Exception {
```

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/graphql/GraphQLClient.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    public <D> D execute(GraphQLOperation operation, Class<D> responseType)
+            throws VisualApiException {
+        try {
+            HttpPost request = new HttpPost(this.uri);
+            request.setHeader(HttpHeaders.CONTENT_TYPE, "application/json");
+            request.setHeader(HttpHeaders.ACCEPT, "application/json");
+            request.setHeader(HttpHeaders.AUTHORIZATION, "Basic " + authentication);
+            request.setHeader(HttpHeaders.USER_AGENT, "sauce-visual-espresso/0.0.1");
```
Comment (konraddysput) [inline visual-espresso/visual/src/main/java/com/saucelabs/visual/graphql/GraphQLClient.java:66]: You can read SDK version via build config

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/junit/TestMetaInfoRule.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                try {
+                    TestMetaInfo.THREAD_LOCAL.set(new TestMetaInfo(className, testName));
+                    base.evaluate();  // Run the test
+                } finally {
+                    TestMetaInfo.THREAD_LOCAL.remove();
+                }
+            }
+        };
+    }
+}
```
Comment (konraddysput) [inline visual-espresso/visual/src/main/java/com/saucelabs/visual/junit/TestMetaInfoRule.java:28]: missing end line in the end of file

**Review by konraddysput — COMMENTED** [review]
Can we please comments in the public API?

**Comment by omacranger** [comment]
A lot of the code here looks similar-ish to the Java client. Can none of that be imported and used here as a dependency somehow?

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/VisualApi.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+public class VisualApi {
+    private final static String LOG_TAG = VisualApi.class.getSimpleName();
+
+    private final GraphQLClient graphQLClient;
+
+    public VisualApi(GraphQLClient graphQLClient) {
+        this.graphQLClient = graphQLClient;
+    }
+
+    VisualBuild createBuild(BuildAttributes buildAttributes) {
```
[the remark this answers is not in the document]
Comment (kb-kerem) [reply visual-espresso/visual/src/main/java/com/saucelabs/visual/VisualApi.java:38]: I understand your concern but leaving out the accessor in java has a special meaning and cannot be expressed with another accessor. And having no accessor here is the most convenient accessor here in this case.
See: https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/VisualApi.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        try {
+            byte[] screenshot = getScreenshot();
+            uploadToUrl(data.result.imageUploadUrl, screenshot);
+        } catch (IOException e) {
+            throw new VisualApiException("uploadSnapshot failed");
+        }
+        return data;
+    }
+
+    private byte[] getScreenshot() throws IOException {
```
[the remark this answers is not in the document]
Comment (kb-kerem) [reply visual-espresso/visual/src/main/java/com/saucelabs/visual/VisualApi.java:69]: 1. makes sense, will do.
2. `os.close()` can cause the IOException

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/VisualBuild.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        return id;
+    }
+
+    private static volatile VisualBuild build = null;
+
+    public static VisualBuild getBuildOnce(VisualApi visualApi, BuildAttributes buildAttributes) {
+        synchronized (VisualBuild.class) {
+            if (build == null) {
+                build = visualApi.createBuild(buildAttributes);
+                Runtime.getRuntime().addShutdownHook(new Thread(() -> visualApi.finishBuild(build.id)));
```
[the remark this answers is not in the document]
Comment (kb-kerem) [reply visual-espresso/visual/src/main/java/com/saucelabs/visual/VisualBuild.java]: The shutdownHook thing doesn't work in Android and this line has been removed altogether

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/junit/TestMetaInfoRule.java (hunk 3 above)
Comment (kb-kerem) [reply visual-espresso/visual/src/main/java/com/saucelabs/visual/junit/TestMetaInfoRule.java:28]: added

**Comment by kb-kerem** [comment]
> A lot of the code here looks similar-ish to the Java client. Can none of that be imported and used here as a dependency somehow?

@omacranger That would be super convenient however it brings also some problems that complicates the implementation. Importing Java client brings transitive dependencies that are not compatible with Android environment, considering the target Android API (apache httpclient, graphql-java etc.). Those dependencies need to be excluded explicitly and need to be maintained for the future versions of java client if new ones show up. Also, since we are using the same package and class names for certain classes that need reimplementation (mainly VisualAPI and GraphQLClient), class/method loading problems arise in runtime. Instead of trying to fix those problems, I'd rather keep the implementations fully separate.

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/espresso/VisualClient.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    public void sauceVisualCheck(String snapshotName, VisualCheckOptions options) {
+        CreateSnapshotUploadMutation.Data data = visualApi.uploadSnapshot(this.build.getId());
+        CreateSnapshotMutation.SnapshotIn input = new CreateSnapshotMutation.SnapshotIn(
+                this.build.getId(),
+                data.result.id,
+                snapshotName,
+                options.resolveTestName(),
+                options.resolveSuiteName(),
+                OperatingSystem.ANDROID,
+                "");
```
Comment (omacranger) [inline visual-espresso/visual/src/main/java/com/saucelabs/visual/espresso/VisualClient.java:84]: Are we able to get OS version here?

File: visual-espresso/visual/build.gradle [hunk 8]
```diff
@@ -0,0 +1,77 @@
+plugins {
+    alias(libs.plugins.android.library)
+    id 'maven-publish'
+}
+
+android {
+    namespace 'com.saucelabs.visual'
```
[the remark this answers is not in the document]
Comment (kb-kerem) [reply visual-espresso/visual/build.gradle:7]: We do not need to change this namespace. Although this looks like a java package name, they're not coupled or related.

graphql-java unfortunately requires minimum Android API level 26 and we're targeting 21. I'll give this one a try and see if it helps: https://www.apollographql.com/docs/kotlin/

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/espresso/VisualClient.java (hunk 7 above)
Comment (kb-kerem) [reply visual-espresso/visual/src/main/java/com/saucelabs/visual/espresso/VisualClient.java]: yes, added

**Review by konraddysput — APPROVED** [review]

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/VisualApi.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            os.write(file);
+            os.flush();
+
+            // Get the response code
+            int responseCode = connection.getResponseCode();
+            if (responseCode != HttpURLConnection.HTTP_OK) {
+                throw new VisualApiException("Snapshot upload failed. Response code: " + responseCode);
+            }
+        } catch (Exception e) {
+            throw new VisualApiException("Snapshot upload failed", e);
```
[the remark this answers is not in the document]
Comment (kb-kerem) [reply visual-espresso/visual/src/main/java/com/saucelabs/visual/VisualApi.java]: Reworked the exception handling, thanks for the tip

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/exception/VisualApiException.java (hunk 1 above)
Comment (kb-kerem) [reply visual-espresso/visual/src/main/java/com/saucelabs/visual/exception/VisualApiException.java]: This doesn't kill the suite but fails the test instead. Same approach we took in java-client. Tested and works fine.

File: visual-espresso/visual/src/androidTest/java/com/saucelabs/visual/SmokeTest.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.saucelabs.visual;
+
+import org.junit.Test;
+
+public class SmokeTest {
+
+    @Test
+    public void check() {
+        VisualClient visual = new VisualClient("Espresso smoke test");
+        visual.sauceVisualCheck("snap");
```
[the remark this answers is not in the document]
Comment (kb-kerem) [reply visual-espresso/visual/src/androidTest/java/com/saucelabs/visual/SmokeTest.java]: done

**Pull request merged**